File: src/test/java/at/searles/parsing/CombinatorTest.java
package at.searles.parsing;

import at.searles.lexer.Lexer;
import at.searles.parsing.printing.ConcreteSyntaxTree;
import at.searles.parsing.printing.EmptyConcreteSyntaxTree;
import at.searles.parsing.printing.PrinterBacktrackException;
import at.searles.regex.RegexParser;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;

public class CombinatorTest {
    private static final Mapping<CharSequence, String> ToString = new Mapping<CharSequence, String>() {
        @Override
        public String parse(ParserStream stream, @NotNull CharSequence left) {
            return left.toString();
        }

        @Nullable
        @Override
        public CharSequence left(@NotNull String result) {
            return result;
        }
    };

    private final Lexer tokenizer = new Lexer();
    private final Parser<String> chr = Parser.fromRegex(RegexParser.parse("[a-z]"), tokenizer, false, ToString);
    private final Recognizer comma = Recognizer.fromString(",", tokenizer, false);
    private final Initializer<String> emptyString = new Initializer<String>() {
        @Override
        public String parse(ParserStream stream) {
            return "";
        }

        @Override
        public ConcreteSyntaxTree print(String s) {
            return s.isEmpty() ? new EmptyConcreteSyntaxTree() : null;
        }
    };
    private final Fold<String, String, String> appendSingleChar = new Fold<String, String, String>() {
        @Override
        public String apply(ParserStream stream, @NotNull String left, @NotNull String right) {
            return left + right;
        }

        @Override
        public String leftInverse(@NotNull String result) {
            return !result.isEmpty() ? result.substring(0, result.length() - 1) : null;
        }

        @Override
        public String rightInverse(@NotNull String result) {
            return !result.isEmpty() ? result.substring(result.length() - 1) : null;
        }

        @Override
        public String toString() {
            return "{append_single_char}";
        }
    };
    private Parser<String> parser;
    private ParserStream input;
    private String parseResult;
    private ConcreteSyntaxTree printResult;
    private boolean error;

    @Before
    public void setUp() {
        error = false;
    }

    // These tests run on a parser that simply parses single chars and
    // attaches these single chars using a simple fold.
    @Test
    public void plus1FailTest() {
        // chr+
        withParser(emptyString.then(Reducer.plus(chr.fold(appendSingleChar))));
        withInput("");
        actParse();

        Assert.assertFalse(error);
        Assert.assertNull(parseResult);
    }

    @Test
    public void plus1SuccessTest() {
        // chr+
        withParser(emptyString.then(Reducer.plus(chr.fold(appendSingleChar))));
        withInput("abc");
        actParse();
        actPrint();

        Assert.assertFalse(error);
        Assert.assertEquals("abc", parseResult);
        Assert.assertEquals("abc", printResult.toString());
    }

    @Test
    public void joinFailTest() {
        // chr+
        withParser(emptyString.then(comma.join(chr.fold(appendSingleChar))));
        withInput("a,,");
        actParse();

        Assert.assertTrue(error);
    }

    @Test
    public void joinSingleCharTest() {
        // chr+
        withParser(emptyString.then(comma.join(chr.fold(appendSingleChar))));
        withInput("a");
        actParse();
        actPrint();

        Assert.assertFalse(error);
        Assert.assertEquals("a", parseResult);
        Assert.assertEquals("a", printResult.toString());
    }

    @Test
    public void joinMultiCharTest() {
        // chr+
        withParser(emptyString.then(comma.join(chr.fold(appendSingleChar))));
        withInput("a,b,c");
        actParse();
        actPrint();

        Assert.assertFalse(error);
        Assert.assertEquals("abc", parseResult);
        Assert.assertEquals("a,b,c", printResult.toString());
    }

    private void actPrint() {
        try {
            printResult = parser.print(parseResult);
        } catch (PrinterBacktrackException ignored) {
        }
    }

    private void actParse() {
        try {
            parseResult = parser.parse(input);
        } catch (ParserLookaheadException ignored) {
            error = true;
        }
    }

    private void withInput(String input) {
        this.input = ParserStream.fromString(input);
    }

    private void withParser(Parser<String> parser) {
        this.parser = parser;
    }
}

File: src/test/java/at/searles/parsing/CombinatorTest.kt
package at.searles.parsing

import at.searles.lexer.Lexer
import at.searles.parsing.Parser.Companion.fromRegex
import at.searles.parsing.Reducer.Companion.rep1
import at.searles.parsing.printing.ConcreteSyntaxTree
import at.searles.parsing.printing.EmptyTree
import at.searles.regexparser.RegexpParser
import org.junit.Assert
import org.junit.Before
import org.junit.Test

class CombinatorTest {
    private val tokenizer = Lexer()
    private val chr = fromRegex(RegexpParser.parse("[a-z]"), tokenizer, ToString)
    private val comma = Recognizer.fromString(",", tokenizer)
    private val emptyString: Initializer<String> = object : Initializer<String> {
        override fun parse(stream: ParserStream): String {
            return ""
        }

        override fun print(item: String): ConcreteSyntaxTree? {
            return if (item.isEmpty()) EmptyTree else null
        }
    }
    private val appendSingleChar: Fold<String, String, String> = object : Fold<String, String, String> {
        override fun apply(stream: ParserStream, left: String, right: String): String {
            return left + right
        }

        override fun leftInverse(result: String): String? {
            return if (result.isNotEmpty()) result.substring(0, result.length - 1) else null
        }

        override fun rightInverse(result: String): String? {
            return if (result.isNotEmpty()) result.substring(result.length - 1) else null
        }

        override fun toString(): String {
            return "{append_single_char}"
        }
    }
    private lateinit var parser: Parser<String>
    private lateinit var input: ParserStream
    private var parseResult: String? = null
    private var printResult: ConcreteSyntaxTree? = null

    @Before
    fun setUp() {
    }

    // These tests run on a parser that simply parses single chars and
    // attaches these single chars using a simple fold.
    @Test
    fun plus1FailTest() {
        // chr+
        withParser(emptyString.plus(chr.plus(appendSingleChar).rep1()))
        withInput("")
        actParse()
        Assert.assertNull(parseResult)
    }

    @Test
    fun plus1SuccessTest() {
        // chr+
        withParser(emptyString.plus(chr.plus(appendSingleChar).rep1()))
        withInput("abc")
        actParse()
        actPrint()
        Assert.assertEquals("abc", parseResult)
        Assert.assertEquals("abc", printResult.toString())
    }

    @Test
    fun joinSingleCharTest() {
        // chr+
        withParser(emptyString.plus(comma.join(chr.plus(appendSingleChar))))
        withInput("a")
        actParse()
        actPrint()
        Assert.assertEquals("a", parseResult)
        Assert.assertEquals("a", printResult.toString())
    }

    @Test
    fun joinMultiCharTest() {
        // chr+
        withParser(emptyString.plus(comma.join(chr.plus(appendSingleChar))))
        withInput("a,b,c")
        actParse()
        actPrint()
        Assert.assertEquals("abc", parseResult)
        Assert.assertEquals("a,b,c", printResult.toString())
    }

    private fun actPrint() {
        printResult = parser.print(parseResult!!)
    }

    private fun actParse() {
//        try {
            parseResult = parser.parse(input)
//        } catch (ignored: BacktrackNotAllowedException) {
//            isError = true
//        }
    }

    private fun withInput(input: String) {
        this.input = ParserStream.create(input)
    }

    private fun withParser(parser: Parser<String>) {
        this.parser = parser
    }

    companion object {
        private val ToString: Mapping<CharSequence, String> = object : Mapping<CharSequence, String> {
            override fun parse(left: CharSequence, stream: ParserStream): String {
                return left.toString()
            }

            override fun left(result: String): CharSequence? {
                return result
            }
        }
    }
}